### PR TITLE
Improved wrapper/tests/documentation for some discrete classes

### PR DESCRIPTION
**Pull request by dellaert**
These are all the cherry-picked commits from #1574, with one additional commit for the new normalization scheme in MPE.

### CODE REVIEW

File: gtsam_unstable/linear/tests/testQPSolver.cpp [hunk 1]
```diff
@@ -190,7 +190,7 @@ TEST(QPSolver, iterate) {
     // do not recompute dual variables after every step!!!
     //    CHECK(assert_equal(expected[it], state.values, 1e-10));
     //    CHECK(assert_equal(expectedDuals[it], state.duals, 1e-10));
-    it++;
+    // it++;
```
Comment (varunagrawal) [inline gtsam_unstable/linear/tests/testQPSolver.cpp:193]: Should we just kill this variable?

File: gtsam/inference/BayesTreeCliqueBase.h [hunk 2]
```diff
@@ -140,9 +140,15 @@ namespace gtsam {
...
 
-    /** is this the root of a Bayes tree ? */
+    /// Return true if this clique is the root of a Bayes tree. 
     inline bool isRoot() const { return parent_.expired(); }
 
+    /// Return the number of children.
+    size_t nrChildren() const { return children.size(); }
+
+    /// Return the child at index i.
+    const derived_ptr operator[](size_t i) const { return children[i]; }
```
Comment (varunagrawal) [inline gtsam/inference/BayesTreeCliqueBase.h:150]: Perhaps make this `children.at(i)` similar to other code in this PR?

**Review by varunagrawal — APPROVED** [review]
LGTM modulo 2 comments

File: gtsam_unstable/linear/tests/testQPSolver.cpp (hunk 1 above)
Comment (dellaert) [reply gtsam_unstable/linear/tests/testQPSolver.cpp:193]: If I do the commented out tests would fail to compile. Opted to comment out.

**Pull request merged**